actions/restore.py

```diff
@@ -0,0 +1,147 @@
+from charmhelpers.core.host import chdir
+from charmhelpers.core.host import service_start
+from charmhelpers.core.host import service_stop
+from shlex import split
+from subprocess import check_call
+from subprocess import check_output
+from subprocess import Popen
+from subprocess import PIPE
+from datetime import datetime
+import hashlib
+import os
+import sys
+import time
+
+opts = layer.options('etcd')
+
+DATESTAMP = datetime.strftime(datetime.now(), '%Y%m%d-%H%M%S')
+ARCHIVE = "etcd-data-{}.tar.gz".format(DATESTAMP)
+
+ETCD_DATA_DIR = opts['etcd_data_dir']
+ETCD_PORT = config('management_port')
+PRIVATE_ADDRESS = unit_get('private-address')
+SKIP_BACKUP = action_get('skip-backup')
+SNAPSHOT_ARCHIVE = resource_get('snapshot')
+TARGET_PATH = action_get('target')
+
+
+def preflight_check():
+    ''' Check preconditions for data restoration '''
+    if not SNAPSHOT_ARCHIVE:
+        action_fail({'result.failed': 'Missing snapshot. See: README.md'})
+        sys.exit(0)
+
+
+def render_backup():
+    ''' Backup existing data in the event of restoration on a dirty unit. '''
+    if not os.path.isdir(ETCD_DATA_DIR):
+        return
+
+    with chdir(ETCD_DATA_DIR):
+        if not SKIP_BACKUP:
+            log('Backing up existing data found in {}'.format(ETCD_DATA_DIR))
+            # tar cvf $TARGET_PATH/$ARCHIVE default
+            archive_path = "{}/{}".format(TARGET_PATH, ARCHIVE)
+            backup_sum = shasum_file(archive_path)
+            action_set({'backup.path': archive_path,
+                        'backup.sha256sum': backup_sum})
+
+
+def unpack_resource():
+    ''' Grab the resource path, and unpack it into $PATH '''
+    cmd = "tar xvf {0} -C {1}".format(SNAPSHOT_ARCHIVE, ETCD_DATA_DIR)
+    check_call(split(cmd))
+
+
+def start_etcd_forked():
+    ''' Start the etcd daemon temporarily to initiate new cluster details '''
+    raw = "/snap/etcd/current/bin/etcd -data-dir={0} -force-new-cluster"
+    cmd = raw.format(ETCD_DATA_DIR)
+    Popen(split(cmd), stdout=PIPE, stderr=PIPE)
+
+
+def pkill_etcd():
+    ''' Kill the temporary forked etcd daemon '''
+    cmd = 'pkill etcd'
+    check_call(split(cmd))
+
+
+def probe_forked_etcd():
+    ''' Block until the forked etcd instance has started and return'''
+    output = b""
+    loop = 0
+    MAX_WAIT = 10
+
+    while b"http://localhost" not in output:
+        try:
+            output = check_output(split('/snap/bin/etcd.etcdctl member list'))
+            loop = loop + 1
+        except:
+            log('Still waiting on forked etcd instance...')
+            output = b""
+            loop = loop + 1
+        time.sleep(1)
+        if loop > MAX_WAIT:
+            raise TimeoutError("Timed out waiting for forked etcd.")
+
+
+def reconfigure_client_advertise():
+    ''' Reconfigure the backup to use host network addresses for client advertise
+        instead of the assumed localhost addressing '''
+    cmd = "/snap/bin/etcd.etcdctl member list"
```

Comment:
Nitpicky, but you use this string twice (other instance on line 87). No objection if you want to define a variable for it.